doc/api/tls.md

```diff
@@ -2,90 +2,138 @@
+Once the CSR file is generated, it can either be sent to a Certificate
+Authority for signing or used to generate a self-signed certificate.
+
+Creating a self-signed certificate using the OpenSSL command-line interface
+is illustrated in the example below:
 
 ```
 openssl x509 -req -in ryans-csr.pem -signkey ryans-key.pem -out ryans-cert.pem
 ```
 
-Alternatively you can send the CSR to a Certificate Authority for signing.
+### Perfect Forward Secrecy
+
+<!-- type=misc -->
+
+The term "[Forward Secrecy]" or "Perfect Forward Secrecy" describes a feature of
+key-agreement (i.e., key-exchange) methods. That is, as opposed to using the
+server and client's private and public key pairs directly to secure the
+TLS/SSL communication channel, the keys are used to negotiate new keys that
+are used specifically and only for the current communication session. 
+Practically, this means that even if the private key of a server or client is 
+compromised, communication can only be decrypted by eavesdroppers if the
+attacker manages to obtain the key-pair specifically generated for the session.
+
+Perfect Forward Secrecy is achieved by randomly generating a key pair for
+key-agreement on every TLS/SSL handshake (in contrast to using the same key for
+all sessions). Methods implementing this technique are called "ephemeral".
+
+Currently two methods are commonly used to achieve Perfect Forward Secrecy (note
+the character "E" appended to the traditional abbreviations):
+
+* [DHE] - An ephemeral version of the Diffie Hellman key-agreement protocol.
+* [ECDHE] - An ephemeral version of the Elliptic Curve Diffie Hellman
+  key-agreement protocol.
+
+Ephemeral methods may have some performance drawbacks, because key generation
+is expensive.
 
-For Perfect Forward Secrecy, it is required to generate Diffie-Hellman
-parameters:
+To use Perfect Forward Secrecy with the `tls` module, it is required to generate
+Diffie-Hellman parameters. The following illustrates the use of the OpenSSL
+command-line interface to generate such parameters:
 
 ```
 openssl dhparam -outform PEM -out dhparam.pem 2048
 ```
 
-To create a .pfx or .p12, do this:
+Once the parameters are generated, those must be used to generate a `.pfx` or
+`.p12` file:
 
 ```
 openssl pkcs12 -export -in agent5-cert.pem -inkey agent5-key.pem \
       -certfile ca-cert.pem -out agent5.pfx
 ```
 
-  - `in`:  certificate
-  - `inkey`: private key
-  - `certfile`: all CA certs concatenated in one file like
-    `cat ca1-cert.pem ca2-cert.pem > ca-cert.pem`
+Where:
+  
+* `in`: is the signed certificate
+* `inkey`: is the associated private key
+* `certfile`: is a concatenation of all Certificate Authority (CA) certs into
+   a single file, e.g. `cat ca1-cert.pem ca2-cert.pem > ca-cert.pem`
 
-## ALPN, NPN and SNI
+### ALPN, NPN and SNI
 
 <!-- type=misc -->
 
 ALPN (Application-Layer Protocol Negotiation Extension), NPN (Next
 Protocol Negotiation) and, SNI (Server Name Indication) are TLS
```

Comment:
Let's add a note about deprecation of NPN. It was never documented officially, and now is slowly going away.